[Kok Bo Chang] iP

C5hives opened the pull request:
### The Lawrence™ Experience

> “You buy candles when you're out of light, but what do you buy when you're out of candles?” – _Unknown_

Need assistance with keeping things in check? **Lawrence** is here to assist you with that. This bot is:
- Text-based
- Easy-ish to learn
- Fast-ish to use

Sneak Peek:
```
public static void main(String[] args) {
    Lawrence primeMinister = new Lawrence(); // it should be the other way round
    primeMinister.run();
}
```

### Installation
Download the .jar file from [here](https://github.com/C5hives/ip/releases/tag/A-Jar).
1. Run it
2. Add your tasks and view them with `list`
3. Let it manage your ~tasks~ country for you!

It's **FREE**! (until the next GST hike) :star_struck:

### Features
- [x] 1.  Managing tasks
- [ ] 2.  Managing deadlines (coming soon)
- [ ] 3.  Reminders (coming soon)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

C5hives committed:
README: Update chatbot name to Lawrence

C5hives committed:
Lawrence.java: Implement skeletal chatbot version

C5hives committed:
Lawrence.java: Add echo feature to chatbot

C5hives committed:
Lawrence.java: Add ability to store & display tasks

C5hives committed:
Task.java: Add new class that allows task tracking

C5hives committed:
Tasks.java: Add collection to track multiple tasks

C5hives committed:
Lawrence.java: Modify class to use Tasks class for tracking

C5hives committed:
Add different task types to be tracked

The original Task class is superseded by the
ToDo, Deadline and Event classes.

Lawrence.java is updated to handle the different
task types accordingly.

C5hives committed:
Add basic automated text UI testing

The single test case emulates user input and verifies
that the Lawrence class responds correctly when fed
a sequence of inputs.

Ideally, more test cases should be incorporated in
the future.

C5hives committed:
Add error messages for user

Most user misinputs will be recognised by the program
instead of crashing unexpectedly.

The test has also been updated to include some faulty
inputs.

C5hives committed:
Add ability to delete existing tasks

C5hives committed:
Add enums to handle different user commands

The enum class is also flexible enough to handle
the capitalisation and partial capitalisation of commands.

C5hives committed:
Add ability to save and read tasks from hard disk

C5hives committed:
Add overloaded constructors to Task implementation

This change is to facilitate initialising tasks that
may already be complete, which will be utilised when
reading from the hard disk.

C5hives committed:
Add statement to close scanner after use

Also renamed io methods for clarity

C5hives committed:
Separate logic for parsing file and user inputs

This will keep parsing logic localised and can be
called accordingly depending on the source of the input

C5hives committed:
Add factory class to handle parse method

The factory will call the relevant class to parse
input strings depending on its source

C5hives committed:
Add option to skip faulty lines when reading file

C5hives committed:
Fix bug from using wrong delimiter to split string

C5hives committed:
Integrate file saving with bot runtime

Also allows bot to initialise existing tasks on startup

C5hives committed:
Merge branch 'branch-Level-7'

C5hives committed:
Convert stored dates and time in LocalDateTime objects

C5hives committed:
Merge branch 'branch-Level-8'

C5hives committed:
Refactor existing classes to be less coupled

C5hives committed:
Refactor command classes

C5hives committed:
Create command factory to handle command creation

C5hives committed:
Merge branch 'oop-refactor'

C5hives committed:
Refactor classes into packages

C5hives committed:
Rename factory classes into parsers

C5hives committed:
Merge branch 'organise-packages'

C5hives committed:
Merge branch 'add-gradle-support'

C5hives committed:
Add gradle support into project

C5hives committed:
Add JUnit tests for CommandParser and Deadline

C5hives committed:
Modify build.gradle to generate jar file

C5hives committed:
Add javadoc comments for Lawrence.java

C5hives committed:
Add javadoc for Command.java

C5hives committed:
Add javadoc for AddTaskCommand.java

C5hives committed:
Add javadoc to CommandType.java

C5hives committed:
Add javadoc for CompleteTaskCommand.java

C5hives committed:
Add javadoc for DeleteTaskCommand.java

C5hives committed:
Add javadoc for DisplayTasksCommand.java

C5hives committed:
Add javadoc to ExitSessionCommand.java

C5hives committed:
Add javadocs for UncompleteTaskCommand.java

C5hives committed:
Add javadoc for TaskFileManager.java

C5hives committed:
Add javadocs for CommandParser.java

C5hives committed:
Add javadocs for classes in the parser package

C5hives committed:
Add javadocs for classes in the task package

C5hives committed:
Add find functionality to list of valid commands

C5hives committed:
Refactor manager and ui to be mutable

C5hives committed:
Refactor to meet coding standard

C5hives committed:
Merge branch 'branch-Level-9'

C5hives committed:
Merge branch 'branch-A-CodingStandard'

hweinian commented on src/main/java/lawrence/app/Lawrence.java:
```diff
@@ -0,0 +1,79 @@
+
+        Scanner sc = new Scanner(System.in);
+        String userInput;
+        boolean shouldContinue = true;
```
Just a nit: consider using a different variable names like isExit

C5hives replied:
`isExit` was used in the example code snippet for the `A-MoreOOP` flag, but I wanted to avoid the double negative that comes with the `while` loop condition check.

To illustrate what I mean:
```
while(!isExit) { // double negative
    // do something
}
```
is a double negative and will take a bit longer for the reader to process the intention of the `while` loop, while
```
while(shouldContinue) {
    // do something
}
```
is much easier to process and understand (in my opinion).

As it stands, using `should` in a Boolean variable is allowed as per the [coding standard](https://se-education.org/guides/conventions/java/index.html#naming). I will keep this unchanged for now, but will keep this in mind and modify things accordingly as the project evolves.

hweinian commented on src/main/java/lawrence/task/TaskList.java:
```diff
@@ -0,0 +1,167 @@
+ * Represents the list of tasks the user has specified to be tracked by the chatbot.
+ */
+public class TaskList {
+    private ArrayList<Task> list;
```
Maybe specify what kind of list? It should also be plural since it is a collection.
e.g. tasklists or tasks

C5hives replied:
Yup, you're right. Will amend this accordingly!

hweinian commented on src/main/java/lawrence/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+ * Represents a real-life task that needs to be completed.
+ */
+public abstract class Task {
+    private boolean complete;
```
boolean variables should sound like booleans. e.g. isComplete

C5hives replied:
Will rename to `isComplete` since there is no need for a getter method `isComplete()` yet.
I had created a forum post to clarify on this, but it seems highly contentious. Implementation seems to be down to personal preference (as long as it is consistent).

See: https://github.com/nus-cs2103-AY2425S1/forum/issues/210

hweinian commented on src/test/java/lawrence/parser/CommandParserTest.java:
```diff
@@ -0,0 +1,138 @@
+
+import static org.junit.jupiter.api.Assertions.*;
+
+public class CommandParserTest {
```
Great job on considering multiple exceptions that could occur! Very extensive test case!

hweinian reviewed (commented):
LGTM! Code is well structured and code follows OOP principles. Adhere to naming convention and proper use of Javadocs. Great job!

ylchin commented on src/main/java/lawrence/command/AddTaskCommand.java:
```diff
@@ -0,0 +1,72 @@
+package lawrence.command;
+
+import lawrence.database.TaskFileManager;
+import lawrence.parser.InputSource;
+import lawrence.parser.TaskParser;
+import lawrence.task.Task;
+import lawrence.task.TaskList;
+import lawrence.ui.UserInterface;
+import lawrence.utils.DateParser;
+
+import java.io.IOException;
+import java.time.format.DateTimeParseException;
```
Wrong import order. Refer to import order as specified in A-CodingStandard. import java.* should come first

C5hives replied:
The import order specified by the [coding standard](https://se-education.org/guides/conventions/java/index.html#package-and-import-statements) only requires that the order of imports is **consistent** across the project, which my project follows. As the original repository does not have any imports in the original `Duke.java` file, there is no import order established prior to my current ordering.

The coding standard's example is merely an example and is not the fixed order that must be followed.

Not a coding standard violation per se, but will refactor to import java packages first as it makes more sense.
Thanks for bringing this up! :)

ylchin commented on src/main/java/lawrence/command/CompleteTaskCommand.java:
```diff
@@ -0,0 +1,64 @@
+package lawrence.command;
+
+import lawrence.database.TaskFileManager;
+import lawrence.task.Task;
+import lawrence.task.TaskList;
+import lawrence.ui.UserInterface;
+
+import java.io.IOException;
```
Same as comment above

C5hives replied:
Same as previous reply

ylchin commented on src/main/java/lawrence/command/FindTasksCommand.java:
LGTM

ylchin commented on src/main/java/lawrence/task/TaskType.java:
LGTM

ylchin reviewed (commented):
Overall, code was quite easy to read. Noted some coding standard violations like naming of booleans and ArrayList, as well as import order

C5hives committed:
Refactor to adhere to coding standard

C5hives committed:
Refactor to adhere to checkstyle guidelines

C5hives committed:
Merge branch 'A-CheckStyle'

C5hives committed:
Add javaFX UI to replace console

C5hives committed:
Add javadocs and ensure code follows checkstyle

C5hives committed:
Merge branch 'add-gui'

C5hives committed:
Refactor manager and ui to be final

C5hives committed:
Merge branch 'branch-Level-10'

The previous branch used to add javafx gui was
add-gui, which was incorrectly named and not
picked up by grading scripts.

The proper branch branch-Level-10 was created
with minimal commits with the sole intent of
getting merged back into the master branch.

Let's merge branch-Level-10 back into master,
but leave the Level-10 tag as is so that the
grading scripts mark the level 10 task as
complete.

C5hives committed:
Modify javadoc for TaskList::findTasks method

The current javadoc description for the
method does not fully explain what instance
fields the query matches when the method
checks for a match.

Changing the javadoc to specify that the
description of the task is used for matching
better explains the purse of the method.

Let's modify the javadoc to specify that
string matching is done with the task
description to improve the documentation
for the function.

C5hives committed:
Add assertions at certain points in program

C5hives committed:
Change behaviour of fromString method

The current implementation attempts to match
the given input with the appriopriate CommandType
enum value. If the input does not match any
enum values, an IllegalArgumentException is thrown.

This behaviour uses an exception for control flow,
which should be avoided.

Adding a special type to signify invalid command
types might be a better way to handle bad inputs.

Let's
- Add an INVALID type for CommandType
- Use the INVALID type as the default return value
of CommandType::fromString

C5hives committed:
Refactor bot's return values after commands

The current implementation for getResponse uses
stores information about the previous executed
command as a workaround for javafx components to
access information about the executed command.

This behaviour contributes to the temporary field
code smell, as additional null checks need to be
made if public methods are not called in the right
order.

Extracting out such information and placing them in
a record class might be better to move data related
to commands.

Lets
- Create a record object Response to store the data
need by JavaFX for rendering
- Refactor command objects to return a string response
in the execute method instead of storing it as an instance
variable to be accessed later
- Clean up other parts of the program that require
said data to use the Response record instead.
- Remove unneeded instance fields that clutter the
classes mentioned

C5hives committed:
Refactor comments to use consistent capitalisation

C5hives committed:
Merge pull request #2 from C5hives/A-Streams

Modify javadoc for TaskList::findTasks method

C5hives committed:
Revert "Merge pull request #2 from C5hives/A-Streams"

This reverts commit 50849baf4fa33c66b3fb9e5e6bce442224c29653, reversing
changes made to 317727a961e3088aa8717a953edea5e56ab2c78e.

C5hives committed:
Rephrase javadocs for TaskList::findTasks

C5hives committed:
Merge branch 'master' into branch-A-Streams

C5hives committed:
Merge pull request #3 from C5hives/branch-A-Streams

Rephrase javadocs for TaskList::findTasks

C5hives committed:
Merge branch 'master' into branch-A-Assertions

C5hives committed:
Merge branch 'master' into branch-A-CodeQuality

C5hives committed:
Merge pull request #4 from C5hives/branch-A-Assertions

Branch a assertions

C5hives committed:
Merge branch 'master' into branch-A-CodeQuality

C5hives committed:
Merge pull request #5 from C5hives/branch-A-CodeQuality

Branch a code quality

C5hives committed:
Add github actions for continuous integration

C5hives committed:
Update workflow.yml for CI

C5hives committed:
Change command type parsing to be flexible

C5hives committed:
Update test case for CommandParser

C5hives committed:
Merge branch 'branch-BCD-Extension'

C5hives committed:
Add window name for chatbot UI

C5hives committed:
Add welcome message on chatbot startup

C5hives committed:
Add screenshot of ui

C5hives committed:
Update README

C5hives committed:
Add config file for github pages

C5hives committed:
Add more examples to README

C5hives committed:
Rename Lawrence instance variable

C5hives committed:
Rename boolean variables in TaskFileManager

C5hives committed:
Something
